### PR TITLE
Revive: dumping/ffmpeg_backend: Various fixes

### PR DESCRIPTION
**Pull request by Steveice10**
Rebase of https://github.com/citra-emu/citra/pull/5411 with an MSVC compile error fixed.

This fixes some of the issues I was seeing in the video dumper with hardware-accelerated codecs. It also fixes codecs that don't support weird timebases by resampling to a normal frame rate.

- - -
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/6528)

### CODE REVIEW

**Comment by FearlessTobi** [comment]
> The dumps should be at the console's native framerate (268111856/4481136 FPS or roughly 59.83122493939037 FPS), not a flat 60 FPS. The latest nightlies are fine in that regard (as in, they dump at 268111856/4481136 FPS).

Has this concern from the original PR been addressed?

File: src/core/dumping/ffmpeg_backend.cpp [hunk 1]
```diff
@@ -102,6 +106,43 @@ FFmpegVideoStream::~FFmpegVideoStream() {
     Free();
 }
 
+// This is modified from libavcodec/decode.c
+// The original version was broken
+static AVPixelFormat GetPixelFormat(AVCodecContext* avctx, const AVPixelFormat* fmt) {
+    // Choose a software pixel format if any, prefering those in the front of the list
+    for (int i = 0; fmt[i] != AV_PIX_FMT_NONE; i++) {
+        const AVPixFmtDescriptor* desc = av_pix_fmt_desc_get(fmt[i]);
+        if (!(desc->flags & AV_PIX_FMT_FLAG_HWACCEL))
```
Comment (FearlessTobi) [inline src/core/dumping/ffmpeg_backend.cpp:115]: please brace this. dito below

File: src/core/dumping/ffmpeg_backend.cpp [hunk 2]
```diff
@@ -198,20 +249,192 @@ void FFmpegVideoStream::ProcessFrame(VideoFrame& frame) {
...
+    if (av_buffersrc_add_frame(source_context, current_frame.get()) < 0) {
+        LOG_ERROR(Render, "Video frame dropped: Could not add frame to filter graph");
         return;
     }
-    if (sws_context) {
-        sws_scale(sws_context.get(), current_frame->data, current_frame->linesize, 0, layout.height,
-                  scaled_frame->data, scaled_frame->linesize);
+    int error = 0;
+    while (true) {
+        error = av_buffersink_get_frame(sink_context, filtered_frame.get());
```
Comment (FearlessTobi) [inline src/core/dumping/ffmpeg_backend.cpp:261]:
```suggestion
       const int error = av_buffersink_get_frame(sink_context, filtered_frame.get());
```

File: src/core/dumping/ffmpeg_backend.cpp [hunk 3]
```diff
@@ -198,20 +249,192 @@ void FFmpegVideoStream::ProcessFrame(VideoFrame& frame) {
...
+        if (error == AVERROR(EAGAIN) || error == AVERROR_EOF)
```
Comment (FearlessTobi) [inline src/core/dumping/ffmpeg_backend.cpp:262]: should be braced

**Comment by Steveice10** [comment]
> > The dumps should be at the console's native framerate ([268111856/4481136](tel:268111856/4481136) FPS or roughly 59.83122493939037 FPS), not a flat 60 FPS. The latest nightlies are fine in that regard (as in, they dump at [268111856/4481136](tel:268111856/4481136) FPS).
> 
> Has this concern from the original PR been addressed?

The 60FPS is intentional, we are using filters to resample from the native frame rate to a solid 60 to provide a simpler timebase, as some codecs do not support the native rate.

**Review by FearlessTobi — COMMENTED** [review]
Looks good to me. We should test this a bit in canary though

File: src/core/dumping/ffmpeg_backend.cpp [hunk 4]
```diff
@@ -125,15 +166,28 @@ bool FFmpegVideoStream::Init(FFmpegMuxer& muxer, const Layout::FramebufferLayout
     codec_context->bit_rate = Settings::values.video_bitrate;
     codec_context->width = layout.width;
     codec_context->height = layout.height;
-    // TODO(xperia64): While these numbers from core timing work fine, certain video codecs do not
-    // support the strange resulting timebase (280071/16756991); Addressing this issue would require
-    // resampling the video
-    // List of codecs known broken by this change: mpeg1, mpeg2, mpeg4, libxvid
-    // See https://github.com/citra-emu/citra/pull/5273#issuecomment-643023325 for more information
-    codec_context->time_base.num = static_cast<int>(GPU::frame_ticks);
```
Comment (kongfl888) [inline src/core/dumping/ffmpeg_backend.cpp:133]: Compared to PR #5411, frame_ticks have been defined as extern after PR #6386. Perhaps it needs to be checked.

File: src/core/dumping/ffmpeg_backend.cpp [hunk 5]
```diff
@@ -198,20 +249,192 @@ void FFmpegVideoStream::ProcessFrame(VideoFrame& frame) {
...
+
+    const AVFilter* source = avfilter_get_by_name("buffer");
+    const AVFilter* sink = avfilter_get_by_name("buffersink");
+    if (!source || !sink) {
+        LOG_ERROR(Render, "Could not find buffer source or sink");
+        return false;
+    }
+
+    // Configure buffer source
+    static constexpr AVRational src_time_base{static_cast<int>(GPU::frame_ticks),
```
Comment (kongfl888) [inline src/core/dumping/ffmpeg_backend.cpp:383]: Compared to PR #5411, frame_ticks have been defined as extern after PR #6386. Perhaps it needs to be checked?

**Pull request merged**